src/oscar/apps/partner/abstract_models.py

```diff
@@ -189,11 +191,27 @@ def allocate(self, quantity):
 
         This normally happens when a product is bought at checkout.  When the
         product is actually shipped, then we 'consume' the allocation.
+
         """
+        # Atomic update
+        (self.__class__.objects
+            .filter(pk=self.pk)
+            .update(num_allocated=(
+                Coalesce(F('num_allocated'), Value(0)) + quantity)))
+
+        # Make sure the current object is up-to-date
         if self.num_allocated is None:
             self.num_allocated = 0
         self.num_allocated += quantity
```

Comment:
We explicitly want to reuse the current value in the database not on the instance because the instance value might be old . See https://docs.djangoproject.com/en/1.10/ref/models/expressions/#avoiding-race-conditions-using-f